src/Authoring/WinRT.SourceGenerator/AotOptimizer.cs

```diff
@@ -0,0 +1,210 @@
+
+            HashSet<ITypeSymbol> compatibleTypesForGeneric = new(SymbolEqualityComparer.Default);
+
+            if (GeneratorHelper.IsWinRTType(type.TypeArguments[0]))
+            {
+                compatibleTypesForGeneric.Add(type.TypeArguments[0]);
+            }
+
+            foreach (var iface in type.TypeArguments[0].AllInterfaces)
+            {
+                if (GeneratorHelper.IsWinRTType(iface))
+                {
+                    compatibleTypesForGeneric.Add(iface);
+                }
+
+                if (iface.IsGenericType
+                    && TryGetCompatibleWindowsRuntimeTypesForVariantType(iface, typeStack, out var compatibleIfaces))
+                {
+                    compatibleTypesForGeneric.UnionWith(compatibleIfaces);
+                }
+            }
+
+            var baseType = type.TypeArguments[0].BaseType;
+            while (baseType != null)
+            {
+                if (GeneratorHelper.IsWinRTType(baseType))
+                {
+                    compatibleTypesForGeneric.Add(baseType);
+                }
+                baseType = baseType.BaseType;
+            }
+
+            typeStack.Pop();
+
+            compatibleTypes = new List<INamedTypeSymbol>(compatibleTypesForGeneric.Count);
+            foreach (var compatibleType in compatibleTypesForGeneric)
+            {
+                compatibleTypes.Add(definition.Construct(compatibleType));
+            }
+
+            return true;
+        }
+
+        private static void GenerateVtableAttributes(SourceProductionContext sourceProductionContext, VtableAttribute vtableAttribute)
+        {
+            // Even if no interfaces are on the vtable, as long as the base type is a WinRT type,
+            // we want to put the attribute as the attribute is the opt-in for the source generated
+            // vtable generation.
+            if (vtableAttribute.Interfaces.Any() || vtableAttribute.HasWinRTExposedBaseType)
+            {
+                var vtableEntries = string.Join(", ", vtableAttribute.Interfaces.Select((@interface) => $@"typeof({@interface})"));
+                
+                StringBuilder source = new();
+                if (!vtableAttribute.IsGlobalNamespace)
+                {
+                    source.Append($@"namespace {vtableAttribute.Namespace}{{");
+                }
+
+                source.Append($@"
+    [global::WinRT.WinRTExposedType({vtableEntries})]
+    partial class {vtableAttribute.ClassName}
+    {{
+    }}");
+
+                if (!vtableAttribute.IsGlobalNamespace)
+                {
+                    source.Append($@"}}");
+                }
+
+                string prefix = vtableAttribute.IsGlobalNamespace ? "" : $"{vtableAttribute.Namespace}.";
+                sourceProductionContext.AddSource($"{prefix}{vtableAttribute.ClassName}.WinRTVtable.g.cs", source.ToString());
+            }
+        }
+    }
+
+    internal readonly struct VtableAttribute

```

Comment:
This model is not equatable, so it'll break the incremental pipeline. You'll need two changes:
- Make this a `sealed record` (it's large enough, not worth making it a struct)
- Replace that `string[]` with `EquatableArray<string>` (copy it from [here](https://github.com/CommunityToolkit/dotnet/blob/main/src/CommunityToolkit.Mvvm.SourceGenerators/Helpers/EquatableArray%7BT%7D.cs))